firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/index.ts

```diff
@@ -100,22 +100,71 @@ export class FirestoreBigQueryEventHistoryTracker
     return data;
   }
 
+  /**
+   * Check whether a failed operation is retryable or not.
+   * Reasons for retrying:
+   * 1) We added a new column to our schema. Sometimes BQ is not ready to stream insertion records immediately
+   *    with the new column.
+   */
+  private async isRetryableInsertionError(e) {
+    let isRetryable = true;
+    const expectedErrors = [
+      {
+        "message": "no such field.",
+        "location": "document_id"
+      }
+    ]
+    if (
+      e.response &&
+      e.response.insertErrors &&
+      e.response.insertErrors.errors
+    ) {
+      const errors = e.response.insertErrors.errors;
+      errors.map((error) => {
+        let isExpected = false;
+        expectedErrors.map(expectedError => {
+          if (error.message === expectedError.message && error.location === expectedError.location) {
+            isExpected = true;
+          }
+        })
+        if (!isExpected) {
+          isRetryable = false;
+        }
+      });
+    }
+    return isRetryable;
+  }
+
   /**
    * Inserts rows of data into the BigQuery raw change log table.
    */
-  private async insertData(rows: bigquery.RowMetadata[]) {
+  private async insertData(
+    rows: bigquery.RowMetadata[],
+    overrideOptions: any,
+    retry: boolean
+  ) {
     const options = {
       skipInvalidRows: false,
       ignoreUnknownValues: false,
       raw: true,
     };
+    for (const key in overrideOptions) {
+      options[key] = overrideOptions[key];
+    }
     try {
       const dataset = this.bq.dataset(this.config.datasetId);
       const table = dataset.table(this.rawChangeLogTableName());
       logs.dataInserting(rows.length);
       await table.insert(rows, options);
       logs.dataInserted(rows.length);
     } catch (e) {
+      console.log("this is error", JSON.stringify(e, null, 2));
+      if (retry && this.isRetryableInsertionError(e)) {
+        retry = false;
+        logs.dataInsertRetried(rows.length);
+        const optionsOverride = { ignoreUnknownValues: true };
+        return this.insertData(rows, optionsOverride, retry);
+      }
```

Comment:
instead of recursing and adding extra arguments to the `insertData` function, could we do this:

```ts
catch (e) {
  console.log("this is error", JSON.stringify(e, null, 2));

  this.initialized = false;

  if (this.isRetryableInsertionError(e)) {
    await table.insert(rows, { ...options, ignoreUnknownValues: true });
    this.initialized = true;
    logs.dataInserted(rows.length);
  } else {
    throw e;
  }
}
```

I haven't checked that `this.initialized` will behave as expected, but something to consider